scylla/src/transport/topology.rs

```diff
@@ -301,12 +400,276 @@ async fn query_keyspaces(conn: &Connection) -> Result<HashMap<String, Keyspace>,
+    // Upon migration from thrift to CQL, Cassandra internally creates a surrogate column "value" of
+    // type EmptyType for dense tables. This resolves into this CQL type name.
+    // This column shouldn't be exposed to the user but is currently exposed in system tables.
+    const THRIFT_EMPTY_TYPE: &str = "empty";
+
+    let mut columns_query = Query::new(
+        "select keyspace_name, table_name, column_name, kind, position, type from system_schema.columns",
+    );
+    columns_query.set_page_size(1024);
+
+    let rows = conn
+        .query_all(&columns_query, &[])
+        .await?
+        .rows
+        .ok_or(QueryError::ProtocolError(
+            "system_schema.columns query response was not Rows",
+        ))?;
+
+    let mut tables_schema = HashMap::with_capacity(rows.len());
+
+    for row in rows.into_typed::<(String, String, String, String, i32, String)>() {
+        let (keyspace_name, table_name, column_name, kind, position, type_) =
+            row.map_err(|_| {
+                QueryError::ProtocolError("system_schema.columns has invalid column type")
+            })?;
+
+        if type_ == THRIFT_EMPTY_TYPE {
+            continue;
+        }
+
+        let entry = tables_schema.entry((keyspace_name, table_name)).or_insert((
+            HashMap::new(), // columns
+            HashMap::new(), // partition key
+            HashMap::new(), // clustering key
+        ));
+
+        let cql_type = map_string_to_cql_type(&type_)?;
+
+        let kind = ColumnKind::from_str(&kind)
+            // FIXME: The correct error type is QueryError:ProtocolError but at the moment it accepts only &'static str
+            .map_err(|_| QueryError::InvalidMessage(format!("invalid column kind {}", kind)))?;
+
+        if kind == ColumnKind::PartitionKey || kind == ColumnKind::Clustering {
+            let key_map = if kind == ColumnKind::PartitionKey {
+                entry.1.borrow_mut()
+            } else {
+                entry.2.borrow_mut()
+            };
+            key_map.insert(position, column_name.clone());
+        }
+
+        entry.0.insert(
+            column_name,
+            Column {
+                type_: cql_type,
+                kind,
+            },
+        );
+    }
+
+    let mut partitioners = query_table_partitioners(conn).await?;
+    let mut result = HashMap::new();
+
+    for ((keyspace_name, table_name), (columns, partition_key_columns, clustering_key_columns)) in
+        tables_schema
+    {
+        let mut partition_key = vec!["".to_string(); partition_key_columns.len()];
+        for (position, column_name) in partition_key_columns {
+            partition_key[position as usize] = column_name;
+        }
+
+        let mut clustering_key = vec!["".to_string(); clustering_key_columns.len()];
+        for (position, column_name) in clustering_key_columns {
+            clustering_key[position as usize] = column_name;
+        }
+
+        result.insert(
```

Comment:
Tip: you can avoid cloning keyspace and table names. You can assign the partitioner to a variable first, then call 

```rust
result.insert((keyspace_name, table_name), Table { columns, parition_key, clustering_key, partition })
```

Reply:
It does not work this way, Rust still complains about the value being moved even after variable assignment :(

Reply:
There is a way out:

```rust
let keyspace_and_table_name = (keyspace_name, table_name);

let partitioner = partitioners
    .remove(&keyspace_and_table_name)
    .unwrap_or_default();

result.insert(
    keyspace_and_table_name,
    Table {
        columns,
        partition_key,
        clustering_key,
        partitioner,
    },
);
```

Reply:
Huh, much simpler than I have expected. I have tried something among those lines but with no success. Thanks :)